src/bitcoin/block.h

```diff
@@ -21,47 +27,137 @@ namespace bitcoin {
-        hashMerkleRoot.SetNull();
-        nTime = 0;
+        hashPrevBlock = BlockHash();
         nBits = 0;
+        vTime.fill(0);
+        nReserved = 0;
         nNonce = 0;
+        nHeaderVersion = 0;
+        vSize.fill(0);
+        nHeight = 0;
+        hashEpochBlock.SetNull();
+        hashMerkleRoot.SetNull();
+        hashExtendedMetadata.SetNull();
     }
 
     bool IsNull() const { return (nBits == 0); }
 
-    uint256 GetHash() const;
+    BlockHash GetHash() const;
+
+    /**
+     * Block time as advertised in the block header.
+     * Might not correspond to actual mined or received time.
+     */
+    int64_t GetBlockTime() const {
+        return uint64_t(vTime[0]) | (uint64_t(vTime[1]) << 8) |
+               (uint64_t(vTime[2]) << 16) | (uint64_t(vTime[3]) << 24) |
+               (uint64_t(vTime[4]) << 32) | (uint64_t(vTime[5]) << 40);
+    }
+
+    /** Set the block time as advertised in the block header. */
+    void SetBlockTime(uint64_t nTime) {
+        vTime = {{
+            uint8_t((nTime & 0x0000000000ff)),
+            uint8_t((nTime & 0x00000000ff00) >> 8),
+            uint8_t((nTime & 0x000000ff0000) >> 16),
+            uint8_t((nTime & 0x0000ff000000) >> 24),
+            uint8_t((nTime & 0x00ff00000000) >> 32),
+            uint8_t((nTime & 0xff0000000000) >> 40),
+        }};
+    }
+
+    /**
+     * Block size, with full encoding, including header, metadata and txs.
+     * This doesn't measure anything, so could diverge from the actual size of
+     * an encoded CBlock.
+     */
+    uint64_t GetSize() const {
+        return uint64_t(vSize[0]) | (uint64_t(vSize[1]) << 8) |
+               (uint64_t(vSize[2]) << 16) | (uint64_t(vSize[3]) << 24) |
+               (uint64_t(vSize[4]) << 32) | (uint64_t(vSize[5]) << 40) |
+               (uint64_t(vSize[6]) << 48);
+    }
+
+    /** Set the advertised block size in the block header. */
+    void SetSize(uint64_t nSize) {
+        vSize = {{
+            uint8_t((nSize & 0x000000000000ff)),
+            uint8_t((nSize & 0x0000000000ff00) >> 8),
+            uint8_t((nSize & 0x00000000ff0000) >> 16),
+            uint8_t((nSize & 0x000000ff000000) >> 24),
+            uint8_t((nSize & 0x0000ff00000000) >> 32),
+            uint8_t((nSize & 0x00ff0000000000) >> 40),
+            uint8_t((nSize & 0xff000000000000) >> 48),
+        }};
+    }
+};
+
+class CBlockMetadataField {
+public:
+    uint32_t nFieldId;
+    std::vector<uint8_t> vData;
 
-    int64_t GetBlockTime() const { return int64_t(nTime); }
+    ADD_SERIALIZE_METHODS
+
+    template <typename Stream, typename Operation>
+       inline void SerializationOp(Stream &s, Operation ser_action) {
+        READWRITE(nFieldId);
+        READWRITE(vData);
+    }
 };
 
 class CBlock : public CBlockHeader {
 public:
-    // network and disk
+    // both for network and disk
+    /** Extended metadata for the block as key-value array */
+    std::vector<CBlockMetadataField> vMetadata;
```

Comment:
I'm not a huge fan of generic untyped key/value byte blob data... this is a potential landmine.. :)  But such discussions perhaps are outside the scope of this PR since clearly you guys decided this was a good idea...